perf/src/bin/perf_client.rs

```diff
@@ -395,33 +497,177 @@ impl RequestStats {
+            bytes,
+            bits_per_second: bytes as f64 * 8.0 / self.period.seconds,
+            sender: stream_stats.sender,
+        })
+    }
+}
+
+impl Serialize for Interval {
+    fn serialize<S>(&self, serializer: S) -> Result<<S as Serializer>::Ok, <S as Serializer>::Error>
+    where
+        S: Serializer,
+    {
+        let mut state = serializer.serialize_struct("Interval", 2)?;
+        state.serialize_field("streams", &self.streams)?;
+        // iperf3 outputs duplicate "sum" entries when run in bidirectional mode
+        // serde does not support duplicate keys, so only output one of the sums
+        if self.send_stream_sum.bytes > 0 {
+            state.serialize_field("sum", &self.send_stream_sum)?;
+        } else {
+            state.serialize_field("sum", &self.recv_stream_sum)?;
+        }
+        state.end()
+    }
+}
+
+#[derive(Copy, Clone, Serialize)]
+struct IntervalPeriod {
+    start: f64,
+    end: f64,
+    seconds: f64,
+}
+
+#[derive(Serialize)]
+struct StreamIntervalStats {
+    id: u64,
+    start: f64,
+    end: f64,
+    seconds: f64,
+    bytes: usize,
+    bits_per_second: f64,
+    sender: bool,
+}
+
+#[derive(Serialize)]
+struct StreamIntervalSumStats {
+    bytes: usize,
+    start: f64,
+    end: f64,
+    seconds: f64,
+    bits_per_second: f64,
+    sender: bool,
+}
+
+impl StreamIntervalSumStats {
+    fn new(period: IntervalPeriod) -> Self {
+        Self {
+            bytes: 0,
+            start: period.start,
+            end: period.end,
+            seconds: period.seconds,
+            bits_per_second: 0.0,
+            sender: false,
+        }
+    }
+
+    fn finish(&mut self) {
+        self.bits_per_second = self.bytes as f64 * 8.0 / self.seconds
+    }
+}
+
+#[derive(Serialize)]
+struct Timestamp {
+    #[serde(serialize_with = "serialize_timestamp")]
+    timestamp: SystemTime,
+}
+
+fn serialize_timestamp<S>(time: &SystemTime, s: S) -> Result<S::Ok, S::Error>
+where
+    S: serde::Serializer,
+{
+    use serde::ser::SerializeMap;
+    let mut state = s.serialize_map(Some(1))?;
+    state.serialize_entry(
+        "timesecs",
+        &time.duration_since(UNIX_EPOCH).unwrap().as_secs(),
+    )?;
+    state.end()
+}
+
+#[derive(Serialize)]
 struct Stats {
     /// Test start time
-    start: Instant,
+    #[serde(skip)]
+    start_instant: Instant,
+    /// Test start system time
+    start: Timestamp,
     /// Durations of complete requests
+    #[serde(skip)]
     duration: Histogram<u64>,
     /// Time from finishing the upload until receiving the first byte of the response
+    #[serde(skip)]
     fbl: Histogram<u64>,
     /// Throughput for uploads
+    #[serde(skip)]
     upload_throughput: Histogram<u64>,
     /// Throughput for downloads
+    #[serde(skip)]
     download_throughput: Histogram<u64>,
     /// The total amount of requests executed
+    #[serde(skip)]
     requests: usize,
     /// The amount of successful requests
+    #[serde(skip)]
     success: usize,
```

Comment:
I wouldn't conflate the serialization model with the recording itself. It also seems like you could compute the histograms on the interval rather than have the streams report it when complete. This would give a better picture of throughput than is currently done.